### PR TITLE
feat: add useTypingAnnounce hook for managing live regions triggered by user input

### PR DESCRIPTION
**Pull request by smhigley**
## Previous Behavior

It's tricky to handle live region messages that are triggered while the user is typing, since firing them immediately risks either interrupting or being immediately interrupted buy the keyboard echo of the user's continued input. The general best practice is to have a debounce based on keyboard input, and fire the live region after ~0.5s of no typing.

We've had a number of partner teams who need this and have either implemented it poorly, or reached out to ask how to do it correctly (e.g. because the live region was never coming through in the screen reader, since the speech queue was always immediately overridden by continued typing).

## New Behavior

This hook builds in that behavior for partner teams without them needing to know or code that specific typing/live region guidance. Instead, they can just call `typingAnnounce('message')` for any notifications triggered by user input.

Uses `announce()` under the hood, so batching and other options are handled by that hook.

### CODE REVIEW

File: packages/react-components/react-aria/library/src/AriaLiveAnnouncer/useDomAnnounce.ts [hunk 1]
```diff
@@ -70,6 +70,7 @@ export const useDomAnnounce_unstable = (): AriaLiveAnnounceFn => {
 
         elementRef.current.innerText = '';
         elementRef.current.appendChild(wrappingEl);
+        console.log('Announcing:', wrappingEl.innerText);
```
Comment (spmonahan) [inline packages/react-components/react-aria/library/src/AriaLiveAnnouncer/useDomAnnounce.ts:73]:
```suggestion

```

File: packages/react-components/react-aria/library/src/useTypingAnnounce/useTypingAnnounce.ts [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+  const [setTypingTimeout, clearTypingTimeout] = useTimeout();
+  const messageQueue = React.useRef<Message[]>([]);
+
+  const callback: MutationCallback = React.useCallback((mutationList, observer) => {
+    clearTypingTimeout();
+    setTypingTimeout(() => {
+      messageQueue.current.forEach(({ message, options }) => {
+        announce(message, options);
+      });
+      messageQueue.current = [];
```
Comment (spmonahan) [inline packages/react-components/react-aria/library/src/useTypingAnnounce/useTypingAnnounce.ts:34]: super nit. Since we're not relying on references to the array for anything, just clear it out rather than create a new array.

```suggestion
      messageQueue.current.length = 0;
```

File: packages/react-components/react-aria/library/src/useTypingAnnounce/useTypingAnnounce.ts [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+
+    if (!observer.current) {
+      // create mutation observer if it doesn't exist
+      observer.current = new win.MutationObserver(callback);
+    }
+
+    observer.current.observe(inputEl.current, valueMutationOptions);
+
+    setTypingTimeout(() => {
+      callback([], observer.current!);
```
Comment (spmonahan) [inline packages/react-components/react-aria/library/src/useTypingAnnounce/useTypingAnnounce.ts:54]: Is this necessary?

```suggestion
      callback([], observer.current);
```

File: packages/react-components/react-aria/stories/src/useTypingAnnounce/useTypingAnnounceContentEditable.stories.tsx [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+    return `You have ${maxLength - count} characters remaining`;
+  } else return undefined;
+};
+
+export const ContentEditable = () => {
+  const inputRef = React.useRef<HTMLSpanElement>(null);
+  const [count, setCount] = React.useState(0);
+  const styles = useStyles();
+
+  // pass typingAnnounce a batchId to ensure new charCount updates override old ones
```
Comment (spmonahan) [inline packages/react-components/react-aria/stories/src/useTypingAnnounce/useTypingAnnounceContentEditable.stories.tsx:34]: This comment seems like it should be on L44 where `typingAnnounce` is called? Just feels out of place here.

File: packages/react-components/react-aria/library/src/useTypingAnnounce/useTypingAnnounce.ts [hunk 5]
```diff
@@ -0,0 +1,59 @@
+import * as React from 'react';
+import { useTimeout } from '@fluentui/react-utilities';
```
Comment (spmonahan) [inline packages/react-components/react-aria/library/src/useTypingAnnounce/useTypingAnnounce.ts:2]: I'm not especially concerned about this here but I've been working with timers where this does matter so I figure I'll mention it and you can decide if it matters or not :)

When the window/tab is backgrounded timers are throttled. The exact throttle amount is browser-specific but it seems to be that they are called at most once per second. Since users are interacting directly with the page when these timers are in play, throttled timers seem like an edge case for this feature but perhaps a quick test is called for just to make sure nothing goes super sideways?

**Review by spmonahan — APPROVED** [review]
Looks good! All my comments are of the nit pick variety :)

File: packages/react-components/react-aria/library/src/useTypingAnnounce/useTypingAnnounce.ts (hunk 5 above)
Comment (smhigley) [reply packages/react-components/react-aria/library/src/useTypingAnnounce/useTypingAnnounce.ts:2]: live regions also should not do anything when a window/tab is in the background, so it should be fine, I believe

File: packages/react-components/react-aria/stories/src/useTypingAnnounce/useTypingAnnounceContentEditable.stories.tsx (hunk 4 above)
Comment (smhigley) [reply packages/react-components/react-aria/stories/src/useTypingAnnounce/useTypingAnnounceContentEditable.stories.tsx:34]: oops, yeah, I moved around some code and looks like I left the comment in the wrong place 😅

File: packages/react-components/react-aria/library/src/AriaLiveAnnouncer/useDomAnnounce.ts (hunk 1 above)
Comment (smhigley) [reply packages/react-components/react-aria/library/src/AriaLiveAnnouncer/useDomAnnounce.ts:73]: 🙈

File: packages/react-components/react-aria/library/src/useTypingAnnounce/useTypingAnnounce.ts (hunk 3 above)
Comment (smhigley) [reply packages/react-components/react-aria/library/src/useTypingAnnounce/useTypingAnnounce.ts:54]: oops, heh, scratch that, I only thought that worked b/c I forgot to also remove the `!`. I added an extra `observer.current` check within the timeout instead.

File: packages/react-components/react-aria/stories/src/useTypingAnnounce/useTypingAnnounceDescription.md [hunk 6]
```diff
@@ -0,0 +1,7 @@
+`useTypingAnnounce` takes a reference to an element (which should be an `input`, `textarea`, or `contenteditable`) and returns a `typingAnnounce` function with the same paramters as the `announce` function returned by `useAnnounce`.
```
Comment (spmonahan) [inline packages/react-components/react-aria/stories/src/useTypingAnnounce/useTypingAnnounceDescription.md:1]: >which should be an `input`, `textarea`, or `contenteditable`

Do you think it's worth enforcing this with a type?

Comment (smhigley) [reply packages/react-components/react-aria/stories/src/useTypingAnnounce/useTypingAnnounceDescription.md:1]: Hm, I think I lean towards not. There might be some other use case for announcing while a user is typing (e.g. Dropdown, where you type to jump to an item, but it isn't actually editable). That's kind of a stretch, but I don't really see a benefit in making it a code restriction.

File: packages/react-components/react-aria/library/src/useTypingAnnounce/useTypingAnnounce.ts [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+};
+
+const valueMutationOptions = {
+  attributes: true,
+  subtree: true,
+  characterData: true,
+  attributeFilter: ['value'],
+};
+
+export const useTypingAnnounce = (inputEl: React.RefObject<HTMLElement>): { typingAnnounce: AriaLiveAnnounceFn } => {
```
Comment (layershifter) [inline packages/react-components/react-aria/library/src/useTypingAnnounce/useTypingAnnounce.ts:19]: Let's change API there to return a ref instead of accepting it i.e.

```
export const useTypingAnnounce = (): { typingAnnounce: AriaLiveAnnounceFn } => {
  const inputRef = React.useRef<HTMLElement>();
}
```

this way we will be able to refactor to callback refs without breaks later if needed

File: packages/react-components/react-aria/library/src/useTypingAnnounce/useTypingAnnounce.ts [hunk 8]
```diff
@@ -0,0 +1,65 @@
...
+      if (!win || !inputEl?.current) {
+        return;
+      }
+
+      messageQueue.current.push({ message, options });
+
+      if (!observer.current) {
+        // create mutation observer if it doesn't exist
+        observer.current = new win.MutationObserver(callback);
+      }
```
Comment (layershifter) [inline packages/react-components/react-aria/library/src/useTypingAnnounce/useTypingAnnounce.ts:53]: Let's move creation of `MutationObserver` to a separate `useEffect()` i.e.

```
React.useEffect(() => {
    const targetWindow = targetDocument?.defaultView;
    if (!targetWindow) {
      return;
    }

  const observer= targetWindow win.MutationObserver(callback);

 observer.current = observer;

  return () => {
        observer.disconnect();
   }
}, [callback])
```

This way we ensure that we there will be issues in concurrent mode

File: packages/react-components/react-aria/library/src/useTypingAnnounce/useTypingAnnounce.ts [hunk 9]
```diff
@@ -0,0 +1,65 @@
...
+        mutationObserver.disconnect();
+      }, 500);
+    },
+    [announce, clearTypingTimeout, setTypingTimeout],
+  );
+
+  const typingAnnounce: AriaLiveAnnounceFn = React.useCallback(
+    (message: string, options: AnnounceOptions = {}) => {
+      const win = targetDocument?.defaultView;
+      if (!win || !inputEl?.current) {
```
Comment (layershifter) [inline packages/react-components/react-aria/library/src/useTypingAnnounce/useTypingAnnounce.ts:44]:
```suggestion
      if (!win || !inputEl.current) {
```

nit

**Review by ValentinaKozlova — APPROVED** [review]

File: packages/react-components/react-aria/library/src/useTypingAnnounce/useTypingAnnounce.ts [hunk 10]
```diff
@@ -0,0 +1,84 @@
...
+  const { announce } = useAnnounce();
+
+  const inputRef = React.useRef<TInputElement>(null);
+  const observer = React.useRef<MutationObserver>();
+  const [setTypingTimeout, clearTypingTimeout] = useTimeout();
+  const messageQueue = React.useRef<Message[]>([]);
+
+  const callback: MutationCallback = React.useCallback(
+    (mutationList, mutationObserver) => {
+      clearTypingTimeout();
```
Comment (layershifter) [inline packages/react-components/react-aria/library/src/useTypingAnnounce/useTypingAnnounce.ts]:
```suggestion
```

AFAIR `useTimeout()` does auto clearing on calling `set*`

File: packages/react-components/react-aria/stories/src/useTypingAnnounce/useTypingAnnounceDefault.stories.tsx [hunk 11]
```diff
@@ -0,0 +1,27 @@
+import * as React from 'react';
+import { AriaLiveAnnouncer, Field, Input, useId } from '@fluentui/react-components';
+import type { InputProps } from '@fluentui/react-components';
+import { useTypingAnnounce } from '../../../library/src/useTypingAnnounce/useTypingAnnounce';
```
Comment (layershifter) [inline packages/react-components/react-aria/stories/src/useTypingAnnounce/useTypingAnnounceDefault.stories.tsx]: Bad import

File: packages/react-components/react-aria/stories/src/useTypingAnnounce/useTypingAnnounceContentEditable.stories.tsx [hunk 12]
```diff
@@ -0,0 +1,65 @@
+import * as React from 'react';
+import { AriaLiveAnnouncer, Field, makeStyles, tokens, useId } from '@fluentui/react-components';
+import { useTypingAnnounce } from '../../../library/src/useTypingAnnounce/useTypingAnnounce';
```
Comment (layershifter) [inline packages/react-components/react-aria/stories/src/useTypingAnnounce/useTypingAnnounceContentEditable.stories.tsx]: Bad import

File: packages/react-components/react-aria/stories/src/useTypingAnnounce/useTypingAnnounceFiltering.stories.tsx [hunk 13]
```diff
@@ -0,0 +1,89 @@
+import * as React from 'react';
+import { AriaLiveAnnouncer, Field, Input, makeStyles, tokens, useId } from '@fluentui/react-components';
+import type { InputProps } from '@fluentui/react-components';
+import { useTypingAnnounce } from '../../../library/src/useTypingAnnounce/useTypingAnnounce';
```
Comment (layershifter) [inline packages/react-components/react-aria/stories/src/useTypingAnnounce/useTypingAnnounceFiltering.stories.tsx]: Bad import

File: packages/react-components/react-aria/stories/src/useTypingAnnounce/index.stories.tsx [hunk 14]
```diff
@@ -0,0 +1,18 @@
+import type { Meta } from '@storybook/react';
+
+import descriptionMd from './useTypingAnnounceDescription.md';
+
+export { Default } from './useTypingAnnounceDefault.stories';
+export { ContentEditable } from './useTypingAnnounceContentEditable.stories';
+export { Filtering } from './useTypingAnnounceFiltering.stories';
```
Comment (layershifter) [inline packages/react-components/react-aria/stories/src/useTypingAnnounce/index.stories.tsx]: AFAIR directories and story files story use PascalCase naming

File: packages/react-components/react-aria/stories/src/useTypingAnnounce/useTypingAnnounceFiltering.stories.tsx (hunk 13 above)
Comment (smhigley) [reply packages/react-components/react-aria/stories/src/useTypingAnnounce/useTypingAnnounceFiltering.stories.tsx]: oops, didn't mean to leave this in as-is, thanks for catching it! 😬
